adapters/lora-vertx/src/main/java/org/eclipse/hono/adapter/lora/impl/LoraProtocolAdapter.java

```diff
@@ -243,6 +269,106 @@ void handleProviderRoute(final HttpContext ctx, final LoraProvider provider) {
         currentSpan.finish();
     }
 
+    private Future<Void> handleCommand(final CommandContext commandContext) {
+        Tags.COMPONENT.set(commandContext.getTracingSpan(), getTypeName());
+        final Sample timer = metrics.startTimer();
+        final Command command = commandContext.getCommand();
+
+        final SubscriptionKey subscriptionKey = new SubscriptionKey(command.getTenant(),
+            Optional.ofNullable(command.getGatewayId()).orElse(command.getDeviceId()));
+        if (!commandSubscriptions.containsKey(subscriptionKey)) {
+            LOG.debug("Received command for unknown gateway {} for tenant {}", subscriptionKey.getDeviceId(),
+                subscriptionKey.getTenant());
+            commandContext.reject(String.format("Received command for unknown gateway {} for tenant {}",
+                subscriptionKey.getDeviceId(), subscriptionKey.getTenant()));
+            return Future.failedFuture(new Exception(String.format("Received command for unknown gateway {} for tenant {}",
+                subscriptionKey.getDeviceId(), subscriptionKey.getTenant())));
+        }
+        final Future<TenantObject> tenantTracker = getTenantConfiguration(subscriptionKey.getTenant(),
+            commandContext.getTracingContext());
+
+        return tenantTracker.compose(tenantObject -> {
+            if (command.isValid()) {
+                return checkMessageLimit(tenantObject, command.getPayloadSize(), commandContext.getTracingContext());
+            } else {
+                return Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_BAD_REQUEST, "malformed command message"));
+            }
+        })
+            .compose(success ->
+                getRegistrationClient().assertRegistration(command.getTenant(), command.getGatewayId(), null,
+                    commandContext.getTracingContext()))
+            .compose(registrationAssertion -> {
+                final CompletableFuture<Boolean> booleanCompletableFuture = onCommandReceived(command, subscriptionKey,
+                    registrationAssertion);
+                final CompletableFuture<Future<Void>> future =
+                    booleanCompletableFuture.handle((aBoolean, throwable) -> {
+                        if (throwable != null) {
+                            return Future.failedFuture(throwable);
+                        }
+                        if (aBoolean) {
+                            return Future.succeededFuture();
+                        }
+                        return Future.failedFuture("Failure while sending command");
+                    });
+                booleanCompletableFuture.join();
+                try {
+                    return future.get();
+                } catch (Exception e) {
+                    return Future.failedFuture(e);
+                }
+            })
+            .onComplete(aVoid -> {
```

Comment:
Should be "onSuccess" here.
(onComplete is also called for a failed Future. This would mean that for a failed Future, both this block and the following "onFailure" block would get called.)